graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,358 @@
+                        writeError(out, id, cause);
+                        graphqlSubscriptions.remove(id);
+                    }
+
+                    @Override
+                    public void complete() {
+                        if (completed) {
+                            return;
+                        }
+                        completed = true;
+                        writeComplete(out, id);
+                        graphqlSubscriptions.remove(id);
+                    }
+                });
+
+        graphqlSubscriptions.put(id, executionResultSubscriber);
+        streamMessage.subscribe(executionResultSubscriber, ctx.eventLoop());
+    }
+
+    private void ensureInitiated() throws Exception {
+        if (!connectionInitiated) {
+            throw new Exception("Connection not initiated");
+        }
+    }
+
+    private String serializeToJson(Object object) throws JsonProcessingException {
+        return mapper.writer().writeValueAsString(object);
+    }
+
+    @Nullable
+    private static String toStringFromJson(String name, @Nullable Object value) {
+        if (value == null) {
+            return null;
+        }
+
+        if (value instanceof String) {
+            return (String) value;
+        } else {
+            throw new IllegalArgumentException("Invalid " + name + " (expected: string)");
+        }
+    }
+
+    /**
+     * This only works reliably if maybeMap is from Json, as maps(objects) in Json
+     * can only have string keys.
+     */
+    @SuppressWarnings("unchecked")
+    private static Map<String, Object> toMapFromJson(@Nullable Object maybeMap) {
+        if (maybeMap == null) {
+            return ImmutableMap.of();
+        }
+
+        if (maybeMap instanceof Map) {
+            final Map<?, ?> map = (Map<?, ?>) maybeMap;
+            if (map.isEmpty()) {
+                return ImmutableMap.of();
+            }
+            return Collections.unmodifiableMap((Map<String, Object>) maybeMap);
+        } else {
+            throw new IllegalArgumentException("Unknown parameter type variables");
+        }
+    }
+
+    private <T> T parseJsonString(String content, TypeReference<T> typeReference)
+            throws JsonProcessingException {
+        return mapper.readValue(content, typeReference);
+    }
+
+    private void writePong(WebSocketWriter out) {
+        out.write("{\"type\":\"pong\"}");
+    }
+
+    private void writeConnectionAck(WebSocketWriter out) {
+        out.write("{\"type\":\"connection_ack\"}");
+    }
+
+    private void writeNext(WebSocketWriter out, String operationId, ExecutionResult executionResult) {
+        final HashMap<String, Object> response = new HashMap<>();
+        response.put("id", operationId);
+        response.put("type", "next");
+        response.put("payload", executionResult.toSpecification());
+        try {
+            final String event = serializeToJson(response);
+            logger.trace("NEXT: {}", event);
+            out.write(event);
```

Comment:
I found that if a connection is closed and a write occurs, an exception can be thrown since the stream `WebSocketWriter` is closed.

```suggestion
            out.tryWrite(event);
```

Can be reproduced by modifying the test:
```
outbound.write("{\"type\":\"connection_init\"}");
outbound.write("{\"type\":\"connection_init\"}");
outbound.write("{\"id\":\"1\",\"type\":\"subscribe\",\"payload\":{\"query\":\"subscription {hello}\"}}");
```